### PR TITLE
hsts: add support for Strict-Transport-Security

### PR DESCRIPTION
**Pull request by bagder**
- [X] enable in the build (configure)
- [X] header parsing
- [X] host name lookup
- [X] unit tests for the above
- [X] CURL_VERSION_HSTS
- [X] CI build
- [X] curl_version_info support
- [X] curl -V output
- [X] curl-config --features
- [X] libcurl API to enable HSTS (`CURLOPT_HSTS_CTRL`)
- [X] command line support for `--hsts`
- [X] `--hsts` docs
- [X] `CURLOPT_HSTS_CTRL` docs
- [X] save cache to disk (and tested in unit test)
- [X] load cache from disk (tested in unit test)
- [X] API to set cache file name (`CURLOPT_HSTS`)
- [X] `CURLOPT_HSTS` docs
- [X] `docs/HSTS.md`
- [X] `--version` docs updated
- [X] adjusted `curl_easy_duphandle` accordingly
- [X] HSTS read/write callbacks 
- [X] man pages for the 4 new setopts for callbacks
- [X] libtest to verify reading/writing HSTS cache with callbacks (1915)

Initial thoughts on API and usage: https://github.com/curl/curl/wiki/HSTS

Targeted to get merged for 7.74.0

### CODE REVIEW

**Comment by JCMais** [comment]
I just found the email talking about this feature on the mailing list and also had a quick look into the wiki.

About this:
```
ability to save to something else than a file
```

Would be incredible if supported from the get-go, we probably could have function just like `READFUNCTION`/`WRITEFUNCTION` but responsible for loading / saving the HSTS cache. These would default to loading the file provided with `CURLOPT_HSTS`.

**Comment by bagder** [comment]
ok, let me write up a proposal for how it could work and run it by you before I start any code on that...

**Comment by bagder** [comment]
@JCMais something like this: https://github.com/curl/curl/wiki/HSTS#possible-future ?

**Comment by JCMais** [comment]
This is great @bagder, thanks for taking the time to look into it.

What would be a scenario that `CURLOPT_HSTSWRITEFUNCTION` would return `CURLSTS_AGAIN`?

**Comment by bagder** [comment]
`CURLSTS_DONE` would be a way for the callback to say "**stop** don't give me any more names", before all names have been passed to it. Basically stopping the write without it being a major error. I'm not sure the distinction is actually important to have.... So it would have to return `CURLSTS_AGAIN` to get called with the next name.

Should the write callback get signaled somehow when the last entry is passed to it?

**Comment by JCMais** [comment]
I see, makes sense.

> Should the write callback get signaled somehow when the last entry is passed to it?

Yes, I think this would be great. Example: to close the stream that is being used to write the data to.

**Comment by bagder** [comment]
I added an index struct to the write callback, and I noticed I had forgot to include the expiry date in the hstsentry struct.

**Comment by bagder** [comment]
A first shot on read/write callbacks is now provided in this PR as well. The wiki description of the funtions is updated to match the code (and there are initial manpages in the branch). This still lacks test cases for the callbacks. I'll appreciate all and any feedback I can get on this.

File: docs/HSTS.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+The `[host name]` is dot-prefixed if it is a includeSubDomain.
+
+The time stamp is when the entry expires.
+
+I considered using wget's file format for the HSTS cache. However, they store the time stamp as the epoch (number of seconds since 1970) and I strongly disagree with using that format. Instead I opted to use a format similar to the curl alt-svc cache file format.
+
+## Possible future additions
+
+ - `CURLOPT_HSTS_PRELOAD` - provide a set of preloaded HSTS host names
+ - ability to save to something else than a file
```
Comment (JCMais) [inline docs/HSTS.md:44]: As the options `CURLOPT_HSTSWRITEFUNCTION` and `CURLOPT_HSTSREADFUNCTION` were added, this line probably can be removed.

File: docs/libcurl/opts/CURLOPT_HSTSREADFUNCTION.3 [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+.SH DESCRIPTION
+Pass a pointer to your callback function, as the prototype shows above.
+
+This callback function gets called by libcurl repeatedly when it populates the
+in-memory HSTS cache.
+
+Set the \fIuserp\fP argument with the \fICURLOPT_HSTSREADDATA(3)\fP option or
+it will be NULL.
+
+When the callback is invoke, the \fIsts\fP pointer points to a pupulated
```
Comment (JCMais) [inline docs/libcurl/opts/CURLOPT_HSTSREADFUNCTION.3:46]:
```suggestion
When the callback is invoked, the \fIsts\fP pointer points to a populated
```

File: docs/libcurl/opts/CURLOPT_HSTSWRITEFUNCTION.3 [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+Pass a pointer to your callback function, as the prototype shows above.
+
+This callback function gets called by libcurl repeatedly to allow the
+application to store the in-memory HSTS cache when libcurl is about to discard
+it.
+
+Set the \fIuserp\fP argument with the \fICURLOPT_HSTSWRITEDATA(3)\fP option or
+it will be NULL.
+
+When the callback is invoked, the \fIsts\fP pointer points to a pupulated
```
Comment (JCMais) [inline docs/libcurl/opts/CURLOPT_HSTSWRITEFUNCTION.3:48]:
```suggestion
When the callback is invoked, the \fIsts\fP pointer points to a populated
```

**Review by JCMais — COMMENTED** [review]
just some minor comments

File: docs/HSTS.md (hunk 1 above)
Comment (bagder) [reply docs/HSTS.md:44]: Indeed!

**Comment by bagder** [comment]
Thanks, addressed now!

**Comment by tlhackque** [comment]
I took a quick look at this proposal.  It appears that it is based on the idea that the HSTS cache is loaded and saved sequentially from a flat file at library initiation and teardown.  I'm not sure this scales - especially if HSTS finally gains widespread traction.

Questions:

- What sets the maximum size of the HSTS cache?  Should that be an option?  What would happen if there were 100M hsts-enabled hosts?[1]  Does the cache lookup work (enough hash buckets?  linear search)?
- What if the backing store was a database?  Shouldn't the read callback include the host being accessed?  (Assuming a lazy init.) 
 This would enable a simple SELECT from a database, or provide a hint to a flat file about what to return.  How would a store operation write back only the dirty (modified) records?  (It's likely that there aren't any...or at least many.  HSTS is long-lived.)
- For many applications (e.g. the curl command line), it seems likely that you only need the records for one or few hosts?  Does it make sense to load the entire cache for this?  OTOH, a persistent library user (e.g. a server) may benefit from a large preload.
- What influences the caching algorithm?  LRU makes sense for random user interactions.  But if you have an application that scans a large number of sites in the same order every time, a preload algorithm would work (and LRU could be counter-productive).

I'm not opposed to a simple flatfile default, but you should consider the alternatives.  (A database doesn't have to be massive - SQLite, BerkeleyDB, or even a purpose-built backend might be used.)

In any case, I suggest thinking about scalablitly and the (implict) use models.  I think that at a minimum, the API should support a database back end, and that the cache load should happen when a cache miss happens, not in library init.  This also includes providing a hostname in an eviction callback, not just "write whatever is left in the cache" at library teardown.

Sorry if I'm missing something - I saw this pop up, but am too time-limited to dive deeply into the code.

Nonetheless, I hope these comments are useful.

[1] Yes, that's a very large number to make the point.  Consider a library user that's a search engine.  Or even 10M or 1M.

**Comment by bagder** [comment]
Thanks for good feedback and excellent questions. I think this also shows that it is a good idea to ship this feature marked as *experimental* to start with, to let users try things out as they are now and we can polish and improve things according to what users want and experience.

> the HSTS cache is loaded and saved sequentially from a flat file at library initiation and teardown

If you use that feature, yes. A library user can deal with it however it likes.

I'm positive most applications using libcurl will not populate the cache with millions of entries even if/when HSTS is used very widely, since very few applications work with that amount of random sites in a browser-like fashion.

If we want to build the system to actually have a **huge** hsts database, then yes we'd need something different. The current design is made with a small dataset in mind.

> What sets the maximum size of the HSTS cache?

Nothing.

> Should that be an option?

Perhaps. But I rather see an actual use for this option before we go ahead and add more bells and whistles.

> Does the cache lookup work (enough hash buckets? linear search)?

There's nothing in neither the file format nor in the API for the callbacks that says anything about how the search is to be done. If someone experiences slow HSTS handling I don't think we will object to performance enhancements. I don't think we need to optimize for such a future hypothetical case now but take it as it comes.

> What if the backing store was a database?

Then you're probably asking for an API that lets the application maintain the database and does the entire lookup (because I can't see a future where libcurl bundles some sort of database). That's not provided here but I wouldn't be against it if someone actually provides a viable use case where this is necessary/beneficial.

We could of course perhaps consider such an "asking API" *instead of* the current load/save API to better allow applications to scale to large amounts of hsts entries.

> For many applications (e.g. the curl command line), it seems likely that you only need the records for one or few hosts? Does it make sense to load the entire cache for this?

What is "the entire cache" for you? I would imagine a typical command line use case have a very small cache of just a bunch of hosts so reading "the entire cache" is very fast and small operation. Separate caches for separate invokes. Having two different curl command line tools using the same file on disk will just make the second overwrite the first with the existing functionality.

> What influences the caching algorithm?

Standard coding practice: first we deliver something basic that works. If there's a need, we improve it over time. If not, basic is fine.

> the cache load should happen when a cache miss happens, not in library init

That's a fair idea. I don't think the current API dictates *when* the cache gets read so we can work on tweaking that.

**Comment by bagder** [comment]
I'll merge this shortly, Buckle up!

**Review by twocountriestrus — COMMENTED** [review]
Muy bueno

**Pull request closed without merging**